server/routerlicious/packages/services/src/dynamodb.ts

```diff
@@ -0,0 +1,491 @@
+        const results = await this.make_query({
+            ExpressionAttributeValues:
+                { ...partition_key.AttributeValuesFragment, ...range_key.AttributeValuesFragment },
+            TableName: this.options.table_name,
+            KeyConditionExpression: `${partition_key.ConditionFragment} and ${range_key.ConditionFragment}`,
+            Limit: 1,
+        }, false);
+
+        const values = (results[0].Items || []).map((item) => { return item.content as T; });
+        return values.length > 0 ? values[0] : null;
+    }
+
+    private compose_range_key_fragment(obj: any): IQueryFragment {
+        const key = this.range_index === UNUSED_RANGE_KEY ? UNUSED_RANGE_KEY : obj[this.range_index] || {};
+        // If this is a ranged query and we have a ranged request
+        if (this.options.limit && isObject(key)) {
+            const range = key as any;
+            const from = range && range.$gt > 0 ?
+                Number(range.$gt) + 1 :
+                1;
+            const to = range && range.$lt > 0 ?
+                Number(range.$lt) - 1 :
+                from + this.options.limit - 1;
+            return {
+                ConditionFragment: "sk between :start and :end",
+                AttributeValuesFragment: { ":start": charwise.encode(from), ":end": charwise.encode(to) },
+            };
+        } else {
+            const Key = isNaN(key) ? key : charwise.encode(key);
+            return {
+                Key,
+                ConditionFragment: "sk = :sk",
+                AttributeValuesFragment: { ":sk": Key },
+            };
+        }
+    }
+
+    private compose_key_fragments(value: any): { partition_key: IQueryFragment, range_key: IQueryFragment } {
+        const partition_key = this.compose_partition_key(value);
+        const range_key = this.compose_range_key_fragment(value);
+        return { partition_key, range_key };
+    }
+
+    private get range_index(): string {
+        return this.options.range_index === undefined ? UNUSED_RANGE_KEY : this.options.range_index;
+    }
+}
+
+export class DynamoDB extends EventEmitter implements core.IDb {
+    constructor(private readonly client: DynamoDBDocument, private readonly options: IDynamoDBOptions) {
+        // e.g https://mongodb.github.io/node-mongodb-native/4.1/classes/MongoClient.html#on event keys
+        super();
+    }
+
+    public async close(): Promise<void> {
+        this.client.destroy();
+        this.emit("close");
+    }
+
+    public collection<T>(name: string): core.ICollection<T> {
+        return new DynamoDBCollection<T>(this.client, this.getCollectionIndexes(name));
+    }
+    /**
+     * Returns predefined options for querying specific collections in dynamodb, consisting of partition / range keys
+     * as well as the table name in which the collection is located
+     * @param name
+     * @returns
+     */
+    private getCollectionIndexes(name: string): IDynamoDBCollectionOptions {
+        switch (name) {
```

Comment:
Hardcoding this information in the database driver might cause problems, when new tables are added. Maybe there is a better central place to put this (ideally in a way that all database drivers can use the information)?